content/blog/eng/from-0-to-infinite-the-final-keyword.md

```diff
@@ -0,0 +1,333 @@
+---
+authors: ["bruce"]
+date: "2018-12-04"
+draft: true
+share: false
+categories: [English, php, oop]
+title: "From 0 to infinite the Final keyword"
+type: "post"
+languageCode: "en-EN"
+toc: true
+---
+
+# Google from zero to infinite
+
+Sometimes I do a full immersion on topics of my interest, generally related to programming, topics on testing, good design, etc. 
+I apply a technique invented by me named: **"Google from zero to infinite"**.
+
+Basically I use one or more keywords in Google, then I literally follow all the links of all the pages. It is a very long and laborious activity. But most of the time I find very interesting documents, otherwise difficult to reach. This time I decided to use the keyword final of PHP.
+
+# The concept behind final keyword
+ 
+The keyword final was introduced in version 5 of php. 
+
+The manual says: 
+
+> Final Keyword. PHP 5 introduces the final keyword, which prevents child classes from overriding a method by prefixing the definition with final. 
+> If the class itself is being defined final then it cannot be extended. Note: Properties cannot be declared final, only classes and methods may be declared as final.
+
+Java has the same keyword, c# too but it is called "sealed" instead of "final", but more or less have the same behavior.
+To prevent a class from being inherited (in our case we will see only this type of use,for example languages like java support further behavior but in different contexts).
+
+First of all we will try to understand what it is for, then we will try to do some additional reasoning, and finally we will talk about of comments on the net.
+
+The concept that expresses the keyword final is that a class can not be extended, the decision must be taken by the programmer in the design stage utilizing just the final keyword. 
+
+There are situations in which extension by inheritance is a good solution, others where it is not possible and others where it is wrong. The keyword final is a "tool" given to the programmer to express this decision.
+
+So when the consumer of the class will meet the keyword final, it will have to use different solutions from the inheritance. And this decision is imposed from the class designer.
+
+Why someone should set such a strong constraint?. The answers are many, all aim of having a robust,  and more manageable code-base. For example to avoid the proliferation of infinite hierarchies of classes such as VCL (see Visual Component Library), which are difficult to extend and maintain. In this case (even being a well-made library) the use of inheritance has been used a lot.
```

Comment:
Maybe we can throw a link to VCL here? Wikipedia?

PS: it's a good way to avoid branching too much on a different topic, and to boost a bit the SEO.

Reply:
Good suggestion, I put this only for remider for me 
https://en.wikipedia.org/wiki/Visual_Component_Library